### PR TITLE
Fix overloads in createAnimatedComponent

### PR DESCRIPTION
**Pull request by Latropos**
## Summary

1. Move the overload with type deprecation to be the last one. It means that if user provides invalid type which isn't covered by any overload he will not get the warning about FlatList deprecation. This warning was very confusing if user wasn't using FlatList at all
2. Add overload of `ComponentType`. Since `ComponentType<P = {}> = ComponentClass<P> | FunctionComponent<P>` we have assume that it is enough to have two overloads (ComponentClass & FunctionComponent), but it isn't

|BEFORE|AFTER|
|--|--|
|Error for FastImage and deprecation for FastImage ❌ | Deprecation for FlatList only 😎|
|Screenshot 2023-12-12 at 13 16 37|image|
|image||

### CODE REVIEW

**Review by m-bert — APPROVED** [review]

**Review by tjzel — CHANGES_REQUESTED** [review]
Can you provide examples in the PR description that show why those changes are necessary?

File: src/createAnimatedComponent/createAnimatedComponent.tsx [hunk 1]
```diff
@@ -87,15 +87,7 @@ type Options<P> = {
...
-/**
- * @deprecated Please use `Animated.FlatList` component instead of calling `Animated.createAnimatedComponent(FlatList)` manually.
- */
-// @ts-ignore This is required to create this overload, since type of createAnimatedComponent is incorrect and doesn't include typeof FlatList
-export function createAnimatedComponent(
-  component: typeof FlatList<unknown>,
-  options?: Options<any>
-): ComponentClass<AnimateProps<FlatListProps<unknown>>>;
-
+// Don't change the order of overloads, since such a change breaks current behavior
```
Comment (piaskowyk) [inline src/createAnimatedComponent/createAnimatedComponent.tsx:90]: It appears to be a flawed approach when our only means of ensuring that we don't break anything is through a comment. Is it possible to implement some TypeScript tests instead? I believe that @tjzel has already worked on something similar.

Comment (Latropos) [reply src/createAnimatedComponent/createAnimatedComponent.tsx:90]: Unfortunately there are no tests for deprecation

File: src/createAnimatedComponent/createAnimatedComponent.tsx [hunk 2]
```diff
@@ -106,6 +98,22 @@ export function createAnimatedComponent<P extends object>(
...
+  options?: Options<P>
+): FunctionComponent<AnimateProps<P>> | ComponentClass<AnimateProps<P>>;
+
+/**
+ * @deprecated Please use `Animated.FlatList` component instead of calling `Animated.createAnimatedComponent(FlatList)` manually.
+ */
+// @ts-ignore This is required to create this overload, since type of createAnimatedComponent is incorrect and doesn't include typeof FlatList
+export function createAnimatedComponent(
+  component: typeof FlatList<unknown>,
+  options?: Options<any>
```
Comment (tjzel) [inline src/createAnimatedComponent/createAnimatedComponent.tsx:114]: Can we get rid of `any` here?

Comment (tjzel) [reply src/createAnimatedComponent/createAnimatedComponent.tsx:114]: I see, let's keep it for now then.

**Review by tjzel — APPROVED** [review]

**Pull request merged**